src/ngx_http_lua_kong_module.c

```diff
@@ -305,3 +305,172 @@ ngx_http_lua_kong_ffi_get_full_client_certificate_chain(ngx_http_request_t *r,
     return NGX_ABORT;
 #endif
 }
+
+
+#if (NGX_HTTP_SSL)
+
+/*
+ * called by ngx_http_upstream_ssl_init_connection right after
+ * ngx_ssl_create_connection to override any parameters in the
+ * ngx_ssl_conn_t before handshake occurs
+ *
+ * c->ssl is guaranteed to be present and valid
+ */
+
+void
+ngx_http_lua_kong_set_upstream_ssl(ngx_http_request_t *r, ngx_connection_t *c)
+{
+    ngx_ssl_conn_t              *sc = c->ssl->connection;
+    ngx_http_lua_kong_ctx_t     *ctx;
+    STACK_OF(X509)              *chain;
+    EVP_PKEY                    *pkey;
+    X509                        *x509;
+#ifdef OPENSSL_IS_BORINGSSL
+    size_t                       i;
+#else
+    int                          i;
+#endif
+
+    ctx = ngx_http_get_module_ctx(r, ngx_http_lua_kong_module);
+    if (ctx == NULL || ctx->upstream_client_certificate_chain == NULL) {
+        ngx_log_debug0(NGX_LOG_DEBUG_HTTP, r->connection->log, 0,
+                       "skip overriding upstream SSL configuration, "
+                       "certificate not set");
+        return;
+    }
+
+    ngx_log_debug0(NGX_LOG_DEBUG_HTTP, c->log, 0,
+                   "overriding upstream SSL configuration");
+
+    chain = ctx->upstream_client_certificate_chain;
+    pkey = ctx->upstream_client_private_key;
+
+    if (sk_X509_num(chain) < 1) {
+        ngx_ssl_error(NGX_LOG_ALERT, c->log, 0,
+                      "invalid client certificate chain provided while "
+                      "handshaking with upstream");
+        goto failed;
+    }
+
+    x509 = sk_X509_value(chain, 0);
+    if (x509 == NULL) {
+        ngx_ssl_error(NGX_LOG_ALERT, c->log, 0, "sk_X509_value() failed");
+        goto failed;
+    }
+
+    if (SSL_use_certificate(sc, x509) == 0) {
+        ngx_ssl_error(NGX_LOG_ALERT, c->log, 0, "SSL_use_certificate() failed");
+        goto failed;
+    }
+
+    /* read rest of the chain */
+
+    for (i = 1; i < sk_X509_num(chain); i++) {
+        x509 = sk_X509_value(chain, i);
+        if (x509 == NULL) {
+            ngx_ssl_error(NGX_LOG_ALERT, c->log, 0, "sk_X509_value() failed");
+            goto failed;
+        }
+
+        if (SSL_add1_chain_cert(sc, x509) == 0) {
+            ngx_ssl_error(NGX_LOG_ALERT, c->log, 0, "SSL_add1_chain_cert() failed");
+            goto failed;
+        }
+    }
+
+    if (SSL_use_PrivateKey(sc, pkey) == 0) {
+        ngx_ssl_error(NGX_LOG_ALERT, c->log, 0, "SSL_use_PrivateKey() failed");
+        goto failed;
+    }
+
+    return;
+
+failed:
+
+    ERR_clear_error();
+}
+
+
+static X509 *
+ngx_http_lua_kong_x509_copy(X509 *in)
+{
+    return X509_up_ref(in) == 0 ? NULL : in;
+}
+
+
+static void
+ngx_http_lua_kong_cleanup(void *data)
+{
+    ngx_http_lua_kong_ctx_t     *ctx = data;
+
+    if (ctx->upstream_client_certificate_chain) {
```

Comment:
An explicit check for `!= NULL` maybe? We do so below in `ngx_http_lua_kong_ffi_set_upstream_client_cert_and_key()`